src/ray/gcs/gcs_server/gcs_resource_scheduler.h

```diff
@@ -0,0 +1,183 @@
+
+ private:
+  /// \brief Calculate one of the resource scores.
+  ///
+  /// \param requested Quantity of one of the required resources.
+  /// \param available Quantity of one of the available resources.
+  /// \return Score of the node.
+  double Calculate(const FractionalResourceQuantity &requested,
+                   const FractionalResourceQuantity &available);
+};
+
+/// Gcs resource scheduler implementation.
+/// Non-thread safe.
+class GcsResourceScheduler {
+ public:
+  GcsResourceScheduler(GcsResourceManager &gcs_resource_manager)
+      : gcs_resource_manager_(gcs_resource_manager),
+        node_scorer_(new LeastResourceScorer()) {}
+
+  virtual ~GcsResourceScheduler() = default;
+
+  /// Schedule the specified resources to the cluster nodes.
+  ///
+  /// \param required_resources_list The resources to be scheduled.
+  /// \param scheduling_type This scheduling strategy.
+  /// \param node_filter_func This function is used to filter candidate nodes. If a node
+  /// returns true, it can be used for scheduling. By default, all nodes in the cluster
+  /// can be used for scheduling.
+  /// \return Scheduling selected nodes, it corresponds to `required_resources_list` one
+  /// by one. If the scheduling fails, an empty vector is returned.
+  std::vector<NodeID> Schedule(
+      const std::vector<ResourceSet> &required_resources_list,
+      const SchedulingType &scheduling_type,
+      const std::function<bool(const NodeID &)> &node_filter_func = nullptr);
+
+ private:
+  /// Filter out candidate nodes which can be used for scheduling.
+  ///
+  /// \param cluster_resources The cluster node resources.
+  /// \param node_filter_func This function is used to filter candidate nodes. If a node
+  /// returns true, it can be used for scheduling. By default, all nodes in the cluster
+  /// can be used for scheduling.
+  /// \return The candidate nodes which can be used for scheduling.
+  absl::flat_hash_set<NodeID> FilterCandidateNodes(
+      const absl::flat_hash_map<NodeID, SchedulingResources> &cluster_resources,
+      const std::function<bool(const NodeID &)> &node_filter_func);
+
+  /// Sort required resources according to the scarcity and capacity of resources.
+  /// We will first schedule scarce resources (such as GPU) and large capacity resources
+  /// to improve the scheduling success rate.
+  ///
+  /// \param required_resources The resources to be scheduled.
+  /// \return The Sorted resources.
+  const std::vector<ResourceSet> &SortRequiredResources(
+      const std::vector<ResourceSet> &required_resources);
+
+  /// Schedule resources according to `STRICT_SPREAD` strategy.
+  ///
+  /// \param required_resources_list The resources to be scheduled.
+  /// \param candidate_nodes The nodes can be used for scheduling.
+  /// \return Scheduling selected nodes, it corresponds to `required_resources_list` one
+  /// by one. If the scheduling fails, an empty vector is returned.
+  std::vector<NodeID> StrictSpreadSchedule(
```

Comment:
Shall we move these schedule strategy functions to different files as well ? both for header & impl. we will have more strategies later and they gonna bloat this file even larger.

Reply:
`NodeScorer` will have more modes, i tend to move it to different files. At present, the scheduling strategy is enough. What is your suggestion?

Reply:
That works. Another thought is, we have quite a few variables that impacts scheduling. `PACK/SPREAD/STRICT_PACK` strategy, `NodeScorer` and `ActorScheduleStrategy`, perhaps we categorize them with more clarity?

Reply:
Gcs resource scheduler support `PACK/SPREAD/STRICT_PACK/STRICT_SPREAD` strategy and use `NodeScorer` to score node. Placement group schedule strategy and Actor schedule strategy will use gcs resource scheduler to choose node for PG and actor. At present, it is clear. :)

Reply:
Thanks for the explanation 👍